### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies & Setup
import pandas as pd
import numpy as np 

# File to Load (Remember to Change These)
purchase_data_file = 'C:/Users/Owner/Desktop/Pandas-challenge/Resources/purchase_data.csv'

# Read File & Store
purchase_data_df = pd.read_csv(purchase_data_file)

# Clean Rows with Missing Data
purchase_data_df.count()
cleaned_data_df = purchase_data_df.dropna(how='any')
cleaned_data_df.count()
purchase_data_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

* Display the total number of players


In [2]:
# Calculate Total Player Count 
total_players = len(purchase_data_df['SN'].value_counts())

# Display Data Frame Summary
player_count = pd.DataFrame({'Total Players': [total_players]}) 
player_count

Total Players
0            576

# Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [3]:
# Calculate Number of Unique Items
unique_items = len(purchase_data_df['Item ID'].unique())

# Calculate Average Price
mean_price = (purchase_data_df['Price'].mean())

# Calculate Total Number of Purchases
total_purchases = (purchase_data_df['Purchase ID'].count())

# Calculate Total Revenue
total_revenue = (purchase_data_df['Price'].sum())

# Display Data Frame Summary
purchasing_analysis_df = pd.DataFrame({'Number of Unique Items': [unique_items],
                                      'Average Price': [mean_price],
                                      'Total Number of Purchases': [total_purchases],
                                        'Total Revenue': [total_revenue]})
purchasing_analysis_df

# Format Price & Round
purchasing_analysis_df.style.format({'Average Price': '${:,.2f}',
                                    'Total Revenue': '${:,.2f}'})

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [4]:
# Purchase Date by Gender
genders_df = purchase_data_df.groupby(['Gender'])

# Count Total SN's by Gender
gender_count = genders_df.nunique()['SN']

# Percentage Value / Total Players
player_percentage = gender_count / total_players * 100

# Display Data Frame Summary
gender_demographics_df = pd.DataFrame({'Percent': player_percentage, 'Count of Gender': gender_count})

# Sort Values & Format Percent
gender_demographics_df.sort_values(["Count of Gender"], ascending = False).style.format({"Percent":"{:.2f}%"})

# Format Column Title Name
gender_demographics_df.index.name = None

gender_demographics_df

Percent  Count of Gender
Female                 14.062500               81
Male                   84.027778              484
Other / Non-Disclosed   1.909722               11


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [5]:
# Calculate Purchase Count
purchase_count = genders_df['Purchase ID'].count()

# Calculate Average Purchase Price
mean_purchase_price = genders_df['Price'].mean()

# Calculate Total Purchase Value by Gender
mean_purchase_total = genders_df['Price'].sum()

# Calculate Average Purchase Total per Person by Gender
average_purchase_per_person = mean_purchase_total / gender_count

# Create Data Frame Summary
gender_analysis = pd.DataFrame({'Purchase Count': purchase_count,
                               'Average Purchase Price': mean_purchase_price,
                               'Total Purchase Value': mean_purchase_total,
                               'Average Purchase Total Per Person': average_purchase_per_person})
# Create Column Title Name
gender_analysis.index.name = "Gender"

# Format Price & Round
gender_analysis.style.format({"Total Purchase Value":"${:,.2f}",
                                  "Average Purchase Price":"${:,.2f}",
                                  "Average Purchase Total Per Person":"${:,.2f}"})

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [6]:
# Establish Bins for Ages
age_bins = [0, 9.99, 14.99, 19.99, 24.99, 29.99, 34.99, 39.99, 500]
labels = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

# Categorize Existing Players by Age Bins
purchase_data_df['Age Group'] = pd.cut(purchase_data_df["Age"], age_bins, labels=labels)

# Group Data by Age Group
age_bin_df = purchase_data_df.groupby('Age Group')

# Categorize by Age Count
age_count_categories = age_bin_df['SN'].nunique()

# Calculate the Numbers & Percentages by Age Group
age_percentage = (age_count_categories / total_players) * 100

# Create Data Frame Summary
age_demographics_df = pd.DataFrame({"Percentage of Players": age_percentage, 'Total Count': age_count_categories})

# Create Column Title Name
age_demographics_df.index.name = None

# Format Percentage & Round 
age_demographics_df.style.format({'Percentage of Players': "{:,.2f}%"})

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [7]:
# Find Purchase Count by Age Group
purchase_count_age = age_bin_df["Purchase ID"].count()

# Find Average Purchase Price by Age Group
mean_purchase_price_age = age_bin_df["Price"].mean()

# Calculate Total Purchase Value by Age Group
total_purchase_value = age_bin_df["Price"].sum()

# Calculate the Average Purchase Price Per Person in the Age Group
mean_purchase_per_person_age = total_purchase_value/age_count_categories

# Create Data Frame Summary
age_demographics = pd.DataFrame({"Purchase Count": purchase_count_age,
                                 "Average Purchase Price": mean_purchase_price_age,
                                 "Total Purchase Value":total_purchase_value,
                                 "Average Purchase Total Per Person": mean_purchase_per_person_age})

# Format Column Title Name
age_demographics.index.name = None

# Format Price & Round
age_demographics.style.format({"Average Purchase Price":"${:,.2f}",
                               "Total Purchase Value":"${:,.2f}",
                               "Average Purchase Total Per Person":"${:,.2f}"})


## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [8]:
# Group Data by SN's
sn_grouped = purchase_data_df.groupby('SN')

# Count Total Purchases
purchase_count_sn = sn_grouped['Purchase ID'].count()

# Calculate Mean by SN's
mean_purchase_price_sn = sn_grouped['Price'].mean()

# Calculate Purchase Total
purchase_total_sn = sn_grouped['Price'].sum()

# Calculate Data Frame (from Above)
top_spenders_df = pd.DataFrame({'Purchase Count': purchase_count_sn, 
                               'Average Purchase Price': mean_purchase_price_sn,
                               'Total Purchase Value': purchase_total_sn})

# Sort Total Purchase Value
formatted_top_spenders_df = top_spenders_df.sort_values(["Total Purchase Value"], ascending=False).head()

# Format & Round Price
formatted_top_spenders_df.style.format({"Average Purchase Total":"${:,.2f}",
                                 "Average Purchase Price":"${:,.2f}", 
                                 "Total Purchase Value":"${:,.2f}"})

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, average item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [9]:
# Create Column Titles
items_df = purchase_data_df[['Item ID', 'Item Name', 'Price']]

# Group by Item ID & Item Name
items_grouped = items_df.groupby(['Item ID', 'Item Name'])

# Calculate for Purchase Count, Average Item Price, and Total Purchase Value
item_purchases = items_grouped['Price'].count()
purchase_value = items_grouped['Price'].sum()
item_price = purchase_value/item_purchases

# Create Data Frame Summary
popular_items_df = pd.DataFrame({'Purchase Count': item_purchases,
                                'Item Price': item_price,
                                'Total Purchase Value': purchase_value})
# Sort Purchase Count Column
popular_items_formatted = popular_items_df.sort_values(["Purchase Count"], ascending=False).head()

# Format & Round Price
popular_items_formatted.style.format({"Item Price":"${:,.2f}",
                                    "Total Purchase Value":"${:,.2f}"})

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [10]:
# Sort by Total Purchase Value
profitable_formatted = popular_items_formatted.sort_values(['Total Purchase Value'], ascending=False).head()

# Format & Round Price
profitable_formatted.style.format({'Item Price': '${:,.2f}',
                                  'Total Purchase Value': '${:,.2f}'})

# Observable Trends Based on the Data

* There are a total of 576 players. The vast majority are male (84%). Notably, there is a much smaller proportion of female players at (14%).
* The biggest group of players are in the 20-24 age range, representing close to about 45% of that age demographic. This age group also makes up the biggest group of spenders followed by the 15-19 year olds.
* The most popular and most profitable at $59.99 was the game Final Critic.